tao_compiler/mlir/disc/transforms/dot_rewriter.cc

```diff
@@ -190,6 +191,469 @@ struct TransposeFoldingConvert : public OpRewritePattern<mhlo::DotGeneralOp> {
+  size_t num_non_contracting_dims = non_contracting_tokens.size();
+  // If a reshape is needed, aka, if num of contracting/non_contracting dims > 1
+  if (num_contracting_dims > 1 || num_non_contracting_dims > 1) {
+    SmallVector<SmallVector<int64_t>> reshape_maps;
+    for (int64_t i = 0; i < batching_tokens.size(); ++i) {
+      SmallVector<int64_t> b({i});
+      reshape_maps.push_back(b);
+    }
+    if (isBNC) {
+      // non_contracting
+      SmallVector<int64_t> n;
+      for (int64_t j = 0; j < non_contracting_tokens.size(); ++j) {
+        n.push_back(batching_tokens.size() + j);
+      }
+      reshape_maps.push_back(n);
+      // contracting
+      SmallVector<int64_t> c;
+      for (int64_t j = 0; j < contracting_tokens.size(); ++j) {
+        c.push_back(batching_tokens.size() + non_contracting_tokens.size() +
+                    j);
+      }
+      reshape_maps.push_back(c);
+    } else {
+      // contracting
+      SmallVector<int64_t> c;
+      for (int64_t j = 0; j < contracting_tokens.size(); ++j) {
+        c.push_back(batching_tokens.size() + non_contracting_tokens.size() +
+                    j);
+      }
+      reshape_maps.push_back(c);
+      // non_contracting
+      SmallVector<int64_t> n;
+      for (int64_t j = 0; j < non_contracting_tokens.size(); ++j) {
+        n.push_back(batching_tokens.size() + j);
+      }
+      reshape_maps.push_back(n);
+    }
+
+    SmallVector<int64_t> reshaped_dims;
+    SmallVector<Value> reshaped_shape_values;
+    auto transposed_shape =
+        result.getType().cast<RankedTensorType>().getShape();
+    for (auto& dims : reshape_maps) {
+      int64_t size_static = 1;
+      Value size_value = rewriter.create<arith::ConstantIndexOp>(loc, 1);
+      for (auto dim : dims) {
+        if (size_static == ShapedType::kDynamicSize ||
+            transposed_shape[dim] == ShapedType::kDynamicSize) {
+          size_static = ShapedType::kDynamicSize;
+        } else {
+          size_static *= transposed_shape[dim];
+        }
+        Value orig_dim_val =
+            transposed_shape[dim] == ShapedType::kDynamicSize
+                ? rewriter.create<tensor::DimOp>(loc, result, dim).getResult()
+                : rewriter
+                      .create<arith::ConstantIndexOp>(loc,
+                                                      transposed_shape[dim])
+                      .getResult();
+        size_value =
+            rewriter.create<arith::MulIOp>(loc, size_value, orig_dim_val);
+      }
+      reshaped_dims.push_back(size_static);
+      reshaped_shape_values.push_back(size_value);
+    }
+    RankedTensorType reshaped_ty =
+        RankedTensorType::get(reshaped_dims, orig_type.getElementType());
+    Value reshaped_shape =
+        rewriter.create<tensor::FromElementsOp>(loc, reshaped_shape_values);
+    result = rewriter.create<mhlo::DynamicReshapeOp>(loc, reshaped_ty, result,
+                                                     reshaped_shape);
```

Comment:
If it is static shape, is there a canonicalization pass to convert dyn-reshape to reshape op?

Reply:
yes, refer to DynamicReshapeOpNotActuallyDynamic in hlo_ops.cc